pkg/migration-tool/planner/planner.go

```diff
@@ -0,0 +1,239 @@
+package planner
+
+import (
+	"context"
+	"fmt"
+	"go.uber.org/atomic"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/schollz/progressbar/v3"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/migration-tool/utils"
+)
+
+var (
+	minute                 = time.Minute.Milliseconds()
+	laIncrement            = minute
+	maxTimeRangeDeltaLimit = minute * 120
+)
+
+// Plan represents the plannings done by the planner.
+type Plan struct {
+	Config
+	Mint int64
+	Maxt int64
+	// Block configs.
+	BlockSizeLimitBytes int64
+	blockCounts         atomic.Int64 // Used in maintaining the ID of the in-memory blocks.
+	pbarMux             *sync.Mutex
+	nextMint            int64
+	lastNumBytes        atomic.Int64
+	lastTimeRangeDelta  int64
+	deltaIncRegion      int64 // Time region for which the time-range delta can continue to increase by laIncrement.
+	// Test configs.
+	Quite         bool   // Avoid progress-bars during logs.
+	TestCheckFunc func() // Helps peek into planner during testing. It is called at createBlock() to check the stats of the last block.
+}
+
+// Config represents configuration for the planner. These properties are meant to be read-only.
+type Config struct {
+	ProgressEnabled    bool
+	JobName            string
+	ProgressMetricURL  string
+	ProgressMetricName string // Name for progress metric.
+}
+
+// InitPlan creates an in-memory planner and initializes it. It is responsible for fetching the last pushed maxt and based on that, updates
+// the mint for the provided migration.
+func Init(plan *Plan) (bool, error) {
+	plan.pbarMux = new(sync.Mutex)
+	plan.blockCounts.Store(0)
+	if plan.Config.ProgressEnabled && plan.Config.ProgressMetricURL == "" {
+		return false, fmt.Errorf("read url for remote-write storage should be provided when progress metric is enabled")
+	}
+	if !plan.Config.ProgressEnabled {
+		log.Info("msg", "Resuming from where we left off is turned off. Starting at the beginning of the provided time-range.")
+	}
+	var found bool
+	if plan.Config.ProgressMetricURL != "" && plan.Config.ProgressEnabled {
+		lastPushedMaxt, found, err := plan.fetchLastPushedMaxt()
+		if err != nil {
+			return false, fmt.Errorf("create plan: %w", err)
+		}
+		if found && lastPushedMaxt > plan.Mint && lastPushedMaxt <= plan.Maxt {
+			plan.Mint = lastPushedMaxt
+			log.Warn("msg", fmt.Sprintf("Resuming from where we left off. Last push was on %d. "+
+				"Resuming from mint: %d to maxt: %d time-range: %d mins", lastPushedMaxt, plan.Mint, plan.Maxt, (plan.Maxt-lastPushedMaxt+1)/time.Minute.Milliseconds()))
+		}
+	}
+	plan.nextMint = plan.Mint
```

Comment:
Let's collect all the progressEnabled code under one if
```suggestion
	if plan.Config.ProgressEnabled { 
	    if plan.Config.ProgressMetricURL == "" {
		return false, fmt.Errorf("read url for remote-write storage should be provided when progress metric is enabled"
             }
		log.Info("msg", "Resuming from where we left off is turned off. Starting at the beginning of the provided time-range.")
	    var found bool
	    lastPushedMaxt, found, err := plan.fetchLastPushedMaxt()
	    if err != nil {
	       return false, fmt.Errorf("create plan: %w", err)
	    }
	    if found && lastPushedMaxt > plan.Mint && lastPushedMaxt <= plan.Maxt {
		plan.Mint = lastPushedMaxt
		log.Warn("msg", fmt.Sprintf("Resuming from where we left off. Last push was on %d. "+
				"Resuming from mint: %d to maxt: %d time-range: %d mins", lastPushedMaxt, plan.Mint, plan.Maxt, (plan.Maxt-lastPushedMaxt+1)/time.Minute.Milliseconds()))
		
	}
	plan.nextMint = plan.Mint
```

Reply:
Ah, this is much better. Thank you very much.